src/memory.c

```diff
@@ -0,0 +1,259 @@
+
+    if (entry == NULL)
+    {
+        return NULL;
+    }
+
+    // Allocate the memory
+    void *mem_ptr = mmap(NULL,
+                         size,
+                         PROT_READ | PROT_WRITE,
+                         MAP_PRIVATE | MAP_ANONYMOUS | MAP_NORESERVE,
+                         -1,
+                         0);
+
+    if (mem_ptr == NULL)
+    {
+        free(entry);
+        return NULL;
+    }
+
+    // Add the memory to vm
+    struct kvm_userspace_memory_region region = { .slot = vm->mem->next_slot,
+                                                  .flags = 0,
+                                                  .guest_phys_addr = phys_addr,
+                                                  .memory_size = size,
+                                                  .userspace_addr =
+                                                      (u64)mem_ptr };
+
+    if (ioctl(vm->vm_fd, KVM_SET_USER_MEMORY_REGION, &region) == -1)
+    {
+        munmap(mem_ptr, size);
+        free(entry);
+        return NULL;
+    }
+
+    entry->guest_phys = phys_addr;
+    entry->memory_ptr = mem_ptr;
+    entry->size = size;
+    entry->slot = vm->mem->next_slot;
+    entry->type = MEMORY_USABLE;
+
+    vm->mem->next_slot += 1;
+
+    return entry;
+}
+
+static struct memory_entry *allocate_mmio(u64 phys, u64 size)
+{
+    struct memory_entry *entry = malloc(sizeof(struct memory_entry));
+
+    if (entry == NULL)
+    {
+        return NULL;
+    }
+
+    entry->guest_phys = phys;
+    entry->memory_ptr = NULL;
+    entry->size = size;
+    entry->slot = 0;
+    entry->type = MEMORY_MMIO;
+
+    return entry;
+}
+
+static s32 check_addr_available(vm_t *vm, u64 phys_addr, u64 size)
+{
+    struct linked_list_elt *current = vm->mem->memory_entries->head;
+
+    while (current != NULL)
+    {
+        struct memory_entry *entry = (struct memory_entry *)current->value;
+
+        if ((entry->guest_phys <= phys_addr
+             && (entry->guest_phys + entry->size) > phys_addr)
+            || ((phys_addr + size) >= entry->guest_phys
+                && phys_addr < entry->guest_phys))
+        {
+            return 0;
+        }
+
+        current = current->next;
+    }
+
+    return 1;
+}
+
+s32 memory_alloc(vm_t *vm, u64 phys_addr, u64 size, u32 type)
+{
+    if (vm == NULL || check_addr_available(vm, phys_addr, size) == 0)
+    {
+        return 0;
+    }
+
+    struct memory_entry *entry = NULL;
+    switch (type)
+    {
+    case MEMORY_USABLE:
+        entry = allocate_usable(vm, phys_addr, size);
+        break;
+    case MEMORY_MMIO:
+        entry = allocate_mmio(phys_addr, size);
+        break;
+    }
+
+    if (entry == NULL)
+    {
+        return 0;
+    }
+
+    linked_list_add(vm->mem->memory_entries, entry);
+
+    return 1;
+}
+
+s64 memory_write(vm_t *vm, u64 dest, u8 *buffer, u64 size)
+{
+    struct memory_entry *entry = find_entry(vm, dest);
+
+    if (entry == NULL)
```

Comment:
Should we check that entry type is `USABLE` ?

Reply:
Done 903776d